src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.cs

```diff
@@ -23,65 +25,239 @@ namespace System.Text.RegularExpressions.Generator
+                    {
+                        Debug.Assert(state is Diagnostic);
+                        return state;
+                    }
+
+                    // If we're unable to generate a full implementation for this regex, report a diagnostic.
+                    // We'll still output a limited implementation that just caches a new Regex(...).
+                    if (!regexType.Method.Tree.Root.SupportsCompilation(out string? reason))
+                    {
+                        return (regexType, reason, Diagnostic.Create(DiagnosticDescriptors.LimitedSourceGeneration, regexType.Method.MethodSyntax.GetLocation()));
+                    }
+
+                    // Generate the core logic for the regex.
+                    Dictionary<string, string[]> requiredHelpers = new();
+                    var sw = new StringWriter();
+                    var writer = new IndentedTextWriter(sw);
+                    writer.Indent += 3;
+                    writer.WriteLine();
+                    EmitRegexDerivedTypeRunnerFactory(writer, regexType.Method, requiredHelpers);
+                    writer.Indent -= 3;
+                    return (regexType, sw.ToString(), requiredHelpers);
                 })
                 .Collect();
 
+            // To avoid invalidating every regex's output when anything from the compilation changes,
+            // we extract from it the only things we care about: whether unsafe code is allowed,
+            // and a name based on the assembly's name, and only that information is then fed into
+            // RegisterSourceOutput along with all of the cached generated data from each regex.
+            IncrementalValueProvider<(bool AllowUnsafe, string? AssemblyName)> compilationDataProvider =
+                context.CompilationProvider
+                .Select((x, _) => (x.Options is CSharpCompilationOptions { AllowUnsafe: true }, x.AssemblyName));
+
             // When there something to output, take all the generated strings and concatenate them to output,
             // and raise all of the created diagnostics.
-            context.RegisterSourceOutput(codeOrDiagnostics, static (context, results) =>
+            context.RegisterSourceOutput(codeOrDiagnostics.Combine(compilationDataProvider), static (context, compilationDataAndResults) =>
             {
-                var code = new List<string>(s_headers.Length + results.Length);
+                ImmutableArray<object?> results = compilationDataAndResults.Left;
+
+                // Any top-level results that are diagnostics indicate failure that should stop compilation.
+                // Report all such diagnostics and then bail if there were any.
+                bool failure = false;
+                foreach (object? result in results)
+                {
+                    if (result is Diagnostic d)
+                    {
+                        context.ReportDiagnostic(d);
+                        failure = true;
```

Comment:
The way this is setup though, if we have a case where there are 10 source generated regexes, and only 1 has a problem, we will stop emitting the remaining 9 that were fine, which will create a lot of more compilation errors. Wouldn't it make more sense to report the diagnostic for the one that  had the problem, but continue emitting the code for the remaining 9 so that the compiler errors are limited to the one regex with the issue?

Reply:
Sure, I can change that.